packages/salesforcedx-vscode-apex/src/commands/apexActionController.ts

```diff
@@ -132,4 +132,173 @@ export class ApexActionController {
+    let namedCredentials;
+    let selectedNamedCredential: string | undefined;
+    let finalNamedCredential: string | undefined;
+    try {
+      const rawQueryData = await (
+        await workspaceContext.getConnection()
+      ).query('SELECT MasterLabel FROM NamedCredential');
+      namedCredentials = {
+        ...rawQueryData,
+        records: this.flattenQueryRecords(rawQueryData.records)
+      };
+    } catch (parseError) {
+      throw new Error(nls.localize('error_parsing_nc'));
+    }
+
+    if (namedCredentials) {
+      const quickPicks = namedCredentials.records.map(nc => nc.MasterLabel as string);
+      quickPicks.push(nls.localize('enter_new_nc'));
+      selectedNamedCredential = await vscode.window.showQuickPick(quickPicks, {
+        placeHolder: nls.localize('select_named_credential')
+      });
+    }
+
+    if (!selectedNamedCredential || selectedNamedCredential === nls.localize('enter_new_nc')) {
+      finalNamedCredential = await vscode.window.showInputBox({
+        prompt: nls.localize('enter_nc_name')
+      });
+    } else {
+      finalNamedCredential = selectedNamedCredential;
+    }
+    return finalNamedCredential;
+  };
+
+  private flattenQueryRecords(rawQueryRecords: JsonMap[]) {
+    // filter out the attributes key
+    return rawQueryRecords.map(({ attributes, ...cleanRecords }) => cleanRecords);
+  }
+
+  private buildESRXml = (
+    existingContent: string | undefined,
+    fullPath: string,
+    namedCredential: string | undefined,
+    oasSpec: string
+  ) => {
+    const baseName = path.basename(fullPath).split('.')[0];
+    const safeOasSpec = oasSpec.replaceAll('"', '&apos;');
+
+    if (existingContent) {
+      const doc = create(existingContent);
+
+      // Correctly update the <schema> tag content
+      const schemaElement = doc.find((node: any) => node.node.nodeName === 'schema');
+      if (schemaElement) {
+        schemaElement.txt(safeOasSpec);
+      } else {
+        throw new Error('The <schema> element was not found in the provided XML.');
+      }
+
+      return doc.end({ prettyPrint: true });
+    }
+
+    const newDoc = create({ version: '1.0', encoding: 'UTF-8' })
+      .ele('ExternalServiceRegistration', { xmlns: 'http://soap.sforce.com/2006/04/metadata' })
+      .ele('description')
+      .txt(`${baseName} External Service`)
+      .up()
+      .ele('label')
+      .txt(baseName)
+      .up()
+      .ele('namedCredentialReference')
+      .txt(namedCredential ?? 'Type here the Named Credential')
+      .up()
+      .ele('registrationProviderType')
+      .txt('Custom')
+      .up()
+      .ele('schema')
+      .txt(safeOasSpec)
+      .up()
+      .ele('schemaType')
+      .txt('OpenApi3')
+      .up()
+      .ele('schemaUploadFileExtension')
+      .txt('yaml')
+      .up()
+      .ele('schemaUploadFileName')
+      .txt(`${baseName}_openapi`)
+      .up()
+      .ele('status')
+      .txt('Complete')
+      .up()
+      .ele('systemVersion')
+      .txt('5')
```

Comment:
According to the ESR [spec](https://developer.salesforce.com/docs/atlas.en-us.api_meta.meta/api_meta/meta_externalserviceregistration.htm?q=external) this should be in the range of 1-3, not sure which but let's start with 3